xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +137,777 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+	if !ok {
+		b.logger.Errorf("received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	// Reject whole config if child policy doesn't exist, don't persist it for
+	// later.
+	bb := balancer.Get(lbCfg.ChildPolicy.Name)
+	if bb == nil {
+		return fmt.Errorf("outlier detection: child balancer %q not registered", lbCfg.ChildPolicy.Name)
+	}
+
+	// It is safe to read b.cfg here without holding the mutex, as the only
+	// write to b.cfg happens later in this function. This function is part of
+	// the balancer.Balancer API, so it is guaranteed to be called in a
+	// synchronous manner, so it cannot race with this read.
+	if b.cfg == nil || b.cfg.ChildPolicy.Name != lbCfg.ChildPolicy.Name {
+		b.childMu.Lock()
+		err := b.child.SwitchTo(bb)
+		if err != nil {
+			b.childMu.Unlock()
+			return err
+		}
+		b.childMu.Unlock()
+	}
+
+	b.mu.Lock()
+	// Inhibit child picker updates until this UpdateClientConnState() call
+	// completes. If needed, a picker update containing the no-op config bit
+	// determined from this config and most recent state from the child will be
+	// sent synchronously upward at the end of this UpdateClientConnState()
+	// call.
+	b.inhibitPickerUpdates = true
+	b.updateUnconditionally = false
+	b.cfg = lbCfg
+
+	addrs := make(map[string]bool, len(s.ResolverState.Addresses))
+	for _, addr := range s.ResolverState.Addresses {
+		addrs[addr.Addr] = true
+		if _, ok := b.addrs[addr.Addr]; !ok {
+			b.addrs[addr.Addr] = newAddressInfo()
+		}
+	}
+	for addr := range b.addrs {
+		if !addrs[addr] {
+			delete(b.addrs, addr)
+		}
+	}
+
+	if b.intervalTimer != nil {
+		b.intervalTimer.Stop()
+	}
+
+	if b.noopConfig() {
+		b.onNoopConfig()
+	} else {
+		b.onIntervalConfig()
+	}
+	b.mu.Unlock()
+
+	b.childMu.Lock()
+	err := b.child.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState:  s.ResolverState,
+		BalancerConfig: b.cfg.ChildPolicy.Config,
+	})
+	b.childMu.Unlock()
+
+	done := make(chan struct{})
+	b.pickerUpdateCh.Put(lbCfgUpdate{
+		lbCfg: lbCfg,
+		done:  done,
+	})
+	<-done
+
+	return err
+}
+
+func (b *outlierDetectionBalancer) ResolverError(err error) {
+	if b.child == nil {
+		return
+	}
+	b.childMu.Lock()
+	defer b.childMu.Unlock()
+	b.child.ResolverError(err)
+}
+
+func (b *outlierDetectionBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	scw, ok := b.scWrappers[sc]
+	if !ok {
+		// Shouldn't happen if passed down a SubConnWrapper to child on SubConn
+		// creation.
+		b.logger.Errorf("UpdateSubConnState called with SubConn that has no corresponding SubConnWrapper")
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(b.scWrappers, scw.SubConn)
+	}
+	b.scUpdateCh.Put(&scUpdate{
+		scw:   scw,
+		state: state,
+	})
+}
+
+func (b *outlierDetectionBalancer) Close() {
+	b.closed.Fire()
+	if b.child != nil {
+		b.childMu.Lock()
```

Comment:
Ok, I see what is happening here. child's operations can't be called asynchronously, so I have this mutex grabbed whenever I call downward. Also, the child can be read outside of the context of the balancer.Balancer API (which we have synchronous guarantees for every operation) when it is Updating the SubConnState downward, since that is handled in run(). Thus, I have the mutex grab protecting that operation atomically, in regards to reading whether if it's nil, and child.UpdateSubConnState(...). I updated the documentation to childMu protects the closing of the child, with the example Easwar had me provide. Let me know if this is sufficient/what you were going for.

Reply:
I could also have it be like childMu helps guarantee other operations on the child aren't called after Close().

Reply:
Maybe we can delete `childMu` and never clear `child`, then?  I think you can just make sure when you do `b.closed.Fire()`, you wait for the `run()` goroutine to exit before calling `b.child.Close()`.  We should theoretically be doing that kind of thing everywhere to ensure all balancer cleanup is done before returning back to the channel.